### PR TITLE
Fine-tune availabilities

### PR DESCRIPTION
**Pull request by devmount**
## Description of the Change

This PR enables users to set custom times for each day of week for their schedule:

image
image

## Benefits

Days can now have custom time limits.

## Applicable Issues

This is the first iteration on #422

### CODE REVIEW

File: frontend/src/components/ScheduleCreation.vue [hunk 1]
```diff
@@ -137,7 +144,9 @@ onMounted(() => {
   generateZoomLink.value = scheduleInput.value.meeting_link_provider === MeetingLinkProviderType.Zoom;
 
   // Set a new reference
-  referenceSchedule.value = { ...scheduleInput.value };
+  // Sidenote: I'm too dumb to properly clone this Object without referencing the same availability array. So for now
+  // this is the good ol' JSON workaround (structuredClone, Object assign or spread won't work here)
+  referenceSchedule.value = JSON.parse(JSON.stringify(scheduleInput.value));
```
Comment (devmount) [inline frontend/src/components/ScheduleCreation.vue:149]: I had a lot of issues due to the availabilities array being referenced from all over the components. The only thing that helped was the JSON workaround. Let me know, if you have a more proper way to deep clone that object 😇

File: frontend/src/elements/AlertBox.vue [hunk 2]
```diff
@@ -123,7 +123,6 @@ const toggleDetails = () => {
   .title {
     font-weight: 700;
     line-height: 1.5;
-    white-space: pre;
```
Comment (devmount) [inline frontend/src/elements/AlertBox.vue:126]: This fixes an issue with long error titles bleeding out of the container.

File: frontend/src/locales/de.json
Comment (devmount) [inline frontend/src/locales/de.json:1]: Sorry, that was me sorting the JSON keys again 😂

File: frontend/src/models.ts [hunk 3]
```diff
@@ -177,12 +177,19 @@ export type ExternalConnectionCollection = {
...
+  schedule_id?: number;
+  day_of_week: number;
+  start_time: string;
+  end_time: string;
+  slot_duration?: number;
 };
 
+export type AvailabilitySet = {
+  [key:number]: Availability[];
+}
```
Comment (devmount) [inline frontend/src/models.ts:191]: I already prepared this for the next step, when we have multiple availability entries per day.

File: frontend/src/tbpro/elements/TextInput.vue [hunk 4]
```diff
@@ -95,6 +96,7 @@ const onChange = () => {
           :maxLength="maxLength"
           @invalid.prevent="onInvalid"
           @change="onChange"
+          @blur="emit('blur')"
```
Comment (devmount) [inline frontend/src/tbpro/elements/TextInput.vue:99]: lmk if there's a better way to bubble those events up

**Review by devmount — COMMENTED** [review]
Some pointers on my changes. Thanks for taking the time to look over them!

File: backend/src/appointment/database/repo/availability.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+from .. import models, schemas, repo
+
+
+""" AVAILABILITY repository functions
+"""
+
+
+def get(db: Session, availability_id: int) -> models.Availability | None:
+    """retrieve availability by id"""
+    if availability_id:
```
Comment (MelissaAutumn) [inline backend/src/appointment/database/repo/availability.py:16]: I'm sure this is just a copy and paste, but since we're typing as an int this will only get called if you pass in 0.

Comment (MelissaAutumn) [reply backend/src/appointment/database/repo/availability.py:16]: Although I guess this makes sense 🤔

File: frontend/src/components/ScheduleCreation.vue (hunk 1 above)
Comment (MelissaAutumn) [reply frontend/src/components/ScheduleCreation.vue:149]: Naw that's fine. Feel free to remove the comments, and just mentioned deep clone.

File: frontend/src/tbpro/elements/TextInput.vue [hunk 6]
```diff
@@ -48,7 +48,7 @@ const props = withDefaults(defineProps<Props>(), {
   maxLength: null,
 });
 
-defineEmits(['submit']);
+const emit = defineEmits(['submit', 'blur']);
 defineExpose({ focus });
```
Comment (MelissaAutumn) [inline frontend/src/tbpro/elements/TextInput.vue:52]: I think you can expose blur
https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/blur

then you can just grab the ref and call .blur. But the event also works.

**Review by MelissaAutumn — COMMENTED** [review]
Thanks! Can you add some tests to make sure custom availability time doesn't break on timezones, and also does the switching weekdays from a list of strings to a list of ints break existing schedules? (like when we do the migration)

File: backend/src/appointment/database/repo/availability.py (hunk 5 above)
Comment (devmount) [reply backend/src/appointment/database/repo/availability.py:16]: This is how we do it in all repo get functions. I thought this is exactly to handle the 0 case, since ids always start at 1... Or maybe I don't understand the problem here 😂

Comment (MelissaAutumn) [reply backend/src/appointment/database/repo/availability.py:16]: just thinking out loud, all good here!

File: frontend/src/tbpro/elements/TextInput.vue (hunk 6 above)
Comment (MelissaAutumn) [reply frontend/src/tbpro/elements/TextInput.vue:52]: (I think the emit is fine)

**Review by rwood-moz — CHANGES_REQUESTED** [review]
Very cool! I pulled the branch and ran the existing E2E tests on desktop Firefox and they pass.

One thing, a corner case. Looks like there's no verification on the end time. Example, when NOT using custom times per day if you attempt to set the availability end time to be earlier than the start time you get this error, as expected:

Screenshot 2025-05-20 at 11 18 54 AM

However if you do the same using `customize per day` it actually saves the setting with no error, and funny thing is, in this case, it creates availability slots on the share link from 9:00am on the Monday all day and all night until 8:00am the next day haha

Screenshot 2025-05-20 at 11 19 25 AM

Screenshot 2025-05-20 at 11 23 58 AM

**Comment by devmount** [comment]
@rwood-moz I added validation now (backend and frontend too). Let me know if you can give it another try.

**Comment by rwood-moz** [comment]
> @rwood-moz I added validation now (backend and frontend too). Let me know if you can give it another try.

Cool thanks for the update! It does show the error now for an end time earlier than the start time on custom days, however it is highlighting the wrong day/field i.e. the invalid time is on the Thursday but it highlights Wednesday:

Screenshot 2025-05-26 at 11 23 36 AM

**Review by rwood-moz — APPROVED** [review]
LGTM (and thanks for the validation fix, latest update worked great)!

**Review by MelissaAutumn — APPROVED** [review]
